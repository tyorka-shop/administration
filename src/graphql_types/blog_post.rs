use async_graphql::{ID, SimpleObject};

#[derive(SimpleObject)]
pub struct BlogPost {
  pub id: ID,
  pub src: String,
  pub url: String,
  pub color: String
}


impl From<&BlogPost> for entity::BlogPost {
  fn from(post: &BlogPost) -> Self {
    Self {
      id: post.id.to_string(),
      src: post.src.clone(),
      url: post.url.clone(),
      color: post.color.clone()
    }
  }
}

impl From<&entity::BlogPost> for BlogPost {
  fn from(post: &entity::BlogPost) -> Self {
    Self {
      id: ID::from(post.id.clone()),
      src: post.src.clone(),
      url: post.url.clone(),
      color: post.color.clone()
    }
  }
}