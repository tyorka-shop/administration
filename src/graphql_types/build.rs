use async_graphql::{SimpleObject, ID};

use super::BuildStatus;


#[derive(Debug, SimpleObject)]
#[graphql(complex)]
pub struct Build {
  pub id: ID,
  pub status: BuildStatus,
  pub date: String,
  #[graphql(skip)]
  pub log: String
}

impl From<entity::Build> for Build {
  fn from(build: entity::Build) -> Self {
    Self {
      id: build.id.into(),
      status: build.status.parse().unwrap(),
      date: build.created_at.to_string(),
      log: build.log
    }
  }
}