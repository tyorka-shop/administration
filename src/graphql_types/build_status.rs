use async_graphql::Enum;

#[derive(Enum, Copy, Clone, Eq, PartialEq, Debug)]
pub enum BuildStatus {
    Pending,
    Done,
    Failure,
}

impl std::str::FromStr for BuildStatus {
  type Err = String;

  fn from_str(s: &str) -> Result<Self, Self::Err> {
      match s {
          "PENDING" => Ok(BuildStatus::Pending),
          "DONE" => Ok(BuildStatus::Done),
          "FAILURE" => Ok(BuildStatus::Failure),
          "FAILED" => Ok(BuildStatus::Failure),
          _ => Err(format!("'{}' is not a valid value for BuildStatus", s)),
      }
  }
}