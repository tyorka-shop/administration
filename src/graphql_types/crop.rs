use image_processing;
use serde::Serialize;

#[derive(Debug, Clone, Serialize, async_graphql::SimpleObject)]
pub struct Point {
    pub x: f64,
    pub y: f64,
}

impl Default for Point {
    fn default() -> Self {
        Self { x: 0.0, y: 0.0 }
    }
}

#[derive(Clone, Debug, Serialize, async_graphql::SimpleObject)]
pub struct Crop {
    pub anchor: Point,

    pub factor: f64,
}

impl Default for Crop {
    fn default() -> Self {
        Self {
            anchor: Point::default(),
            factor: 100.0,
        }
    }
}

impl From<image_processing::Crop> for Crop {
    fn from(crop: image_processing::Crop) -> Self {
        Self {
            anchor: Point {
                x: crop.anchor.x,
                y: crop.anchor.y,
            },
            factor: crop.factor,
        }
    }
}
