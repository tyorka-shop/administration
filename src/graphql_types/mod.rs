mod picture_size;
mod crop;
mod picture;
mod product;
mod multi_lang;
mod blog_post;
mod user;


pub use product::{Product, ProductInput};
pub use picture::Picture;
pub use crop::Crop;
pub use picture_size::PictureSize;
pub use blog_post::BlogPost;
pub use multi_lang::MultiLang;
pub use user::User;

