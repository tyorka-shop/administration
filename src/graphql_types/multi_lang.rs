use async_graphql::{InputObject, SimpleObject};
use serde::Serialize;


#[derive(SimpleObject, InputObject, Serialize, Clone)]
#[graphql(input_name = "MultiLangInput")] 
pub struct MultiLang {
  pub en: String,
  pub ru: String
}

impl MultiLang {
  pub fn to_text(&self) -> Self {
    Self { 
      en: markdown_to_text::convert(&self.en),
      ru: markdown_to_text::convert(&self.ru)
    }
  }

  pub fn to_html(&self) -> Self {
    Self { 
      en: markdown::to_html(&self.en),
      ru: markdown::to_html(&self.ru)
    }
  }
}