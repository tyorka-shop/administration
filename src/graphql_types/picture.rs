use super::{
    crop::{Crop, Point},
    picture_size::PictureSize,
};
use async_graphql::ID;
use serde::Serialize;

#[derive(Debug, Serialize, async_graphql::SimpleObject)]
pub struct Picture {
    pub id: ID,
    pub color: String,
    pub original_size: PictureSize,
    pub crop: Crop,
    #[serde(skip_serializing)]
    #[graphql(skip)]
    pub product_id: Option<String>,
}

impl From<entity::Picture> for Picture {
    fn from(row: entity::Picture) -> Self {
        Self {
            id: ID::from(row.id),
            color: row.color,
            original_size: PictureSize {
                width: row.original_size_width,
                height: row.original_size_height,
            },
            crop: Crop {
                anchor: Point {
                    x: row.crop_anchor_x,
                    y: row.crop_anchor_y,
                },
                factor: row.crop_factor,
            },
            product_id: row.product_id,
        }
    }
}

impl Picture {
    pub fn new(filename: &str, width: i64, height: i64, dominant_color: &str, crop: &Crop) -> Self {
        Self {
            id: ID::from(filename),
            color: dominant_color.to_string(),
            original_size: PictureSize { width, height },
            crop: crop.clone(),
            product_id: None,
        }
    }
}

impl From<&Picture> for entity::Picture {
    fn from(pic: &Picture) -> Self {
        Self {
            id: pic.id.to_string(),
            color: pic.color.clone(),
            original_size_width: pic.original_size.width,
            original_size_height: pic.original_size.height,
            crop_anchor_x: pic.crop.anchor.x,
            crop_anchor_y: pic.crop.anchor.y,
            crop_factor: pic.crop.factor,
            product_id: pic.product_id.clone(),
            idx: None,
        }
    }
}

impl Picture {
    pub async fn get_by_product_id(db: &sqlx::SqlitePool, id: &ID) -> sqlx::Result<Vec<Self>> {
        let rows = sqlx::query_as!(
            entity::Picture,
            "select * from pictures where product_id = ? order by `idx`",
            id.0
        )
        .fetch_all(db)
        .await
        .unwrap()
        .into_iter()
        .map(|row| row.into())
        .collect::<Vec<Self>>();

        Ok(rows)
    }
}
