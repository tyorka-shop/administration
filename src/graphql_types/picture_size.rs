use serde::Serialize;

#[derive(Debug, Serialize, async_graphql::SimpleObject)]
pub struct PictureSize {
  pub width: i64,
  pub height: i64,
}