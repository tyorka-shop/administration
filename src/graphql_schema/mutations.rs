use async_graphql::{Context, Object, Result, ID};
use sqlx::SqlitePool;

use crate::{
    graphql_types::{Crop, Picture, Product, ProductInput},
    image_storage::ImageStorage,
};

pub struct Mutations;

#[Object]
impl Mutations {
    async fn save_product<'a>(&self, ctx: &Context<'a>, product: ProductInput) -> Result<Product> {
        let db = ctx.data::<SqlitePool>().unwrap();

        let entity = entity::Product::from(&product);

        if product.pictures.len() == 0 {
            return Err("Product must have at least one picture".into());
        }

        if product.show_in_shop {
            match product.price {
                Some(value) if value <= 0 => {
                    return Err("Product price must be greater than 0".into())
                }
                None => return Err("Product price must be set".into()),
                _ => {}
            }
        }

        entity.insert_or_update(db).await?;

        for (i, picture_id) in product.pictures.iter().enumerate() {
            let index = i.to_string();
            let product_id = product.id.to_string();
            let picture_id = picture_id.to_string();
            sqlx::query!(
                "update pictures set product_id = $1, `idx` = $2 where id = $3",
                product_id,
                index,
                picture_id
            )
            .execute(db)
            .await?;
        }

        Ok(Product::from(&entity))
    }

    async fn save_crop(&self, ctx: &Context<'_>, id: ID, crop: Crop) -> Result<Picture> {
        let db = ctx.data::<SqlitePool>().unwrap();
        let images = ctx.data::<ImageStorage>().unwrap();

        let row = entity::Picture::get_by_id(&db, &id).await.unwrap();

        images.recrop(&row.id, &crop.clone().into()).unwrap();

        let mut pic = Picture::from(row);
        pic.crop = crop.into();
        entity::Picture::from(&pic)
            .insert_or_update(db)
            .await
            .unwrap();

        Ok(pic)
    }

    async fn save_gallery_order(&self, ctx: &Context<'_>, list: Vec<ID>) -> Result<Vec<Product>> {
        let db = ctx.data::<SqlitePool>().unwrap();

        let list = list.iter().map(|id| id.to_string()).collect();

        entity::Product::save_gallery_order(&db, &list)
            .await
            .unwrap();

        let products = entity::Product::get_gallery(&db)
            .await
            .unwrap()
            .iter()
            .map(Product::from)
            .collect();

        Ok(products)
    }

    async fn save_shop_order(&self, ctx: &Context<'_>, list: Vec<ID>) -> Result<Vec<Product>> {
        let db = ctx.data::<SqlitePool>().unwrap();

        let list = list.iter().map(|id| id.to_string()).collect();

        entity::Product::save_shop_order(&db, &list).await.unwrap();

        let products = entity::Product::get_shop(&db)
            .await
            .unwrap()
            .iter()
            .map(Product::from)
            .collect();

        Ok(products)
    }
}

#[cfg(test)]
mod save_product {
    use crate::graphql_types::ProductInput;

    use super::Mutations;
    use async_graphql::{EmptySubscription, InputType, Object, Request, Result, Variables};

    struct Queries;

    #[Object]
    impl Queries {
        async fn status(&self) -> Result<String> {
            Ok("Ok".into())
        }
    }

    fn make_request(product: &ProductInput) -> Request {
        let vars = Variables::from_json(serde_json::json!({ "product": product.to_value() }));

        let mutation = r#"mutation AddProduct($product: ProductInput!) { saveProduct(product: $product) { id } }"#;

        Request::new(mutation).variables(vars)
    }

    async fn request(product: &ProductInput) -> (async_graphql::Response, sqlx::SqlitePool) {
        let schema = async_graphql::Schema::build(Queries, Mutations, EmptySubscription).finish();
        let db = crate::test_utils::setup_db().await.unwrap();

        let request = make_request(product);

        (schema.execute(request.data(db.clone())).await, db)
    }

    #[tokio::test]
    pub async fn valid() {
        let product = ProductInput::new_fixture();
        let (result, _) = request(&product).await;

        assert_eq!(result.errors.first(), None);

        insta::assert_json_snapshot!(result.data.into_json().unwrap());
    }

    #[tokio::test]
    pub async fn empty_pictures() {
        let mut product = ProductInput::new_fixture();
        product.pictures = vec![];

        let (result, _) = request(&product).await;

        assert_eq!(
            result.errors.first().unwrap().message,
            "Product must have at least one picture"
        );
    }

    #[tokio::test]
    pub async fn price_must_be_set() {
        let mut product = ProductInput::new_fixture();
        product.show_in_shop = true;
        product.price = None;

        let (result, _) = request(&product).await;

        assert_eq!(
            result.errors.first().unwrap().message,
            "Product price must be set"
        );
    }

    #[tokio::test]
    pub async fn price_must_be_gt_0() {
        let mut product = ProductInput::new_fixture();
        product.show_in_shop = true;
        product.price = Some(0);

        let (result, _) = request(&product).await;

        assert_eq!(
            result.errors.first().unwrap().message,
            "Product price must be greater than 0"
        );
    }

    #[tokio::test]
    pub async fn update() {
        let schema = async_graphql::Schema::build(Queries, Mutations, EmptySubscription).finish();
        let db = crate::test_utils::setup_db().await.unwrap();

        let mut product = ProductInput::new_fixture();

        let request = make_request(&product).data(db.clone());

        // create
        schema.execute(request).await;

        product.title.en = "New title".into();
        let request = make_request(&product).data(db.clone());

        // update
        let result = schema.execute(request).await;

        assert_eq!(
            serde_json::to_string(&result).unwrap(),
            serde_json::json!({"data":{"saveProduct": {"id": product.id}}}).to_string()
        );

        let row = sqlx::query!("select * from products where id = ? ", product.id.0)
            .fetch_one(&db)
            .await
            .unwrap();

        assert_eq!(row.title_en, "New title");
    }
}

#[cfg(test)]
mod entity_order {
    use async_graphql::{EmptySubscription, Object, Request, Response, Result, Variables};
    use sqlx::SqlitePool;

    use crate::guard::Role;

    use super::Mutations;

    struct Queries;

    #[Object]
    impl Queries {
        async fn status(&self) -> Result<String> {
            Ok("Ok".into())
        }
    }

    async fn request(request: Request, db: &SqlitePool) -> Response {
        let schema = async_graphql::Schema::build(Queries, Mutations, EmptySubscription).finish();

        schema
            .execute(
                request
                    .data(db.clone())
                    .data(Role::Admin("admin".to_string())),
            )
            .await
    }

    async fn save_gallery_order(db: &SqlitePool, list: Vec<String>) -> Result<Response> {
        let vars = Variables::from_json(serde_json::json!({ "list": list }));
        let r = Request::new(
            r#"
            mutation SaveGalleryOrder($list: [ID!]!) {
                saveGalleryOrder(list: $list) {
                    id
                }
            }
            "#,
        )
        .variables(vars);

        let response = request(r, db).await;
        assert_eq!(response.errors.first(), None);

        Ok(response)
    }

    async fn save_shop_order(db: &SqlitePool, list: Vec<String>) -> Result<Response> {
        let vars = Variables::from_json(serde_json::json!({ "list": list }));
        let r = Request::new(
            r#"
            mutation SaveShopOrder($list: [ID!]!) {
                saveShopOrder(list: $list) {
                    id
                }
            }
            "#,
        )
        .variables(vars);

        let response = request(r, db).await;
        assert_eq!(response.errors.first(), None);

        Ok(response)
    }

    async fn setup() -> Result<SqlitePool> {
        let db = crate::test_utils::setup_db().await.unwrap();

        for i in 1..3 {
            let id = i.to_string();
            let mut product = entity::Product::new_fixture_with_id(&id);
            product.show_in_gallery = true;
            product.show_in_shop = true;
            product.price = Some(100);
            product.insert(&db).await.unwrap();
        }

        Ok(db)
    }

    #[derive(serde::Serialize)]
    struct OrderEntity {
        pub entity_id: String,
        pub idx: i64,
    }

    async fn get_entity_order(db: &SqlitePool) -> Result<Vec<OrderEntity>> {
        let rows = sqlx::query_as!(
            OrderEntity,
            "select `entity_id`, `idx` from `entity_order` order by `idx` asc"
        )
        .fetch_all(db)
        .await?;

        Ok(rows)
    }

    #[tokio::test]
    async fn gallery_db_asc() {
        let db = setup().await.unwrap();

        save_gallery_order(&db, vec!["1".into(), "2".into()])
            .await
            .unwrap();

        let rows = get_entity_order(&db).await.unwrap();
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0].entity_id, "1");
        assert_eq!(rows[1].entity_id, "2");
    }

    #[tokio::test]
    async fn gallery_response_asc() {
        let db = setup().await.unwrap();

        let response = save_gallery_order(&db, vec!["1".into(), "2".into()])
            .await
            .unwrap();
        insta::assert_json_snapshot!(response.data.into_json().unwrap());
    }

    #[tokio::test]
    async fn gallery_db_desc() {
        let db = setup().await.unwrap();

        save_gallery_order(&db, vec!["2".into(), "1".into()])
            .await
            .unwrap();

        let rows = get_entity_order(&db).await.unwrap();
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0].entity_id, "2");
        assert_eq!(rows[1].entity_id, "1");
    }

    #[tokio::test]
    async fn gallery_response_desc() {
        let db = setup().await.unwrap();

        let response = save_gallery_order(&db, vec!["2".into(), "1".into()])
            .await
            .unwrap();

        assert_eq!(response.errors.len(), 0);
        insta::assert_json_snapshot!(response.data.into_json().unwrap());
    }

    #[tokio::test]
    async fn shop_db_asc() {
        let db = setup().await.unwrap();

        save_shop_order(&db, vec!["1".into(), "2".into()])
            .await
            .unwrap();

        let rows = get_entity_order(&db).await.unwrap();
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0].entity_id, "1");
        assert_eq!(rows[1].entity_id, "2");
    }

    #[tokio::test]
    async fn shop_response_asc() {
        let db = setup().await.unwrap();

        let response = save_shop_order(&db, vec!["1".into(), "2".into()])
            .await
            .unwrap();
        insta::assert_json_snapshot!(response.data.into_json().unwrap());
    }

    #[tokio::test]
    async fn shop_db_desc() {
        let db = setup().await.unwrap();

        save_shop_order(&db, vec!["2".into(), "1".into()])
            .await
            .unwrap();

        let rows = get_entity_order(&db).await.unwrap();
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0].entity_id, "2");
        assert_eq!(rows[1].entity_id, "1");
    }

    #[tokio::test]
    async fn shop_response_desc() {
        let db = setup().await.unwrap();

        let response = save_gallery_order(&db, vec!["2".into(), "1".into()])
            .await
            .unwrap();

        assert_eq!(response.errors.len(), 0);
        insta::assert_json_snapshot!(response.data.into_json().unwrap());
    }
}
