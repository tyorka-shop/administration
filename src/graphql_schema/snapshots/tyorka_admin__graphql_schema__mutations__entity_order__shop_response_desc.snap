---
source: src/graphql_schema/mutations.rs
expression: response.data.into_json().unwrap()
---
{
  "saveGalleryOrder": [
    {
      "id": "2"
    },
    {
      "id": "1"
    }
  ]
}
